source/jetstream/iterator.go

```diff
@@ -0,0 +1,270 @@
+// Copyright © 2022 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package jetstream
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	sdk "github.com/conduitio/conduit-connector-sdk"
+	"github.com/nats-io/nats.go"
+)
+
+// heartbeatTimeout is a default heartbeat timeout for push consumers.
+const heartbeatTimeout = 2 * time.Second
+
+// Iterator is a iterator for JetStream communication model.
+// It receives message from NATS JetStream.
+type Iterator struct {
+	sync.Mutex
+
+	conn          *nats.Conn
+	messages      chan *nats.Msg
+	unackMessages []*nats.Msg
+	jetstream     nats.JetStreamContext
+	consumerInfo  *nats.ConsumerInfo
+	subscription  *nats.Subscription
+}
+
+// IteratorParams contains incoming params for the NewIterator function.
+type IteratorParams struct {
+	Conn          *nats.Conn
+	BufferSize    int
+	Durable       string
+	Stream        string
+	Subject       string
+	SDKPosition   sdk.Position
+	DeliverPolicy nats.DeliverPolicy
+	AckPolicy     nats.AckPolicy
+}
+
+// NewIterator creates new instance of the Iterator.
+func NewIterator(ctx context.Context, params IteratorParams) (*Iterator, error) {
+	jetstream, err := params.Conn.JetStream()
+	if err != nil {
+		return nil, fmt.Errorf("get jetstream context: %w", err)
+	}
+
+	consumerOpts, err := getConsumerOptions(params)
+	if err != nil {
+		return nil, fmt.Errorf("get consumer options: %w", err)
+	}
+
+	messages := make(chan *nats.Msg, params.BufferSize)
+	subscription, err := jetstream.ChanSubscribe(params.Subject, messages, consumerOpts...)
```

Comment:
Similar question to the one we had in NATS Pubsub: can we handle async errors?

Reply:
Added the async errors handler in the https://github.com/conduitio-labs/conduit-connector-nats-jetstream/pull/5/commits/d0a67bbc82040d3011e993ba3470348ede079d94